rfcs/0181-list-index-syntax.md

```diff
@@ -0,0 +1,290 @@
+It is, in my opinion, unlikely that list and string manipulation (assuming that any feature in competition for this syntax would involve lists or strings somehow) would be so common in Nix to make this a compelling objection.
+
+# Alternatives
+[alternatives]: #alternatives
+
+#### `expr.${0}`
+
+The simplest alternative, aside from doing nothing, would be to reuse the `.${expr}` syntax that Nix already supports in attrpaths.
+In this alternative, when evaluating `expr.${idx}`, the interpreter would determine whether `idx` evaluates to an integer or a string.
+`expr` would be required to be a list if the former and an attribute set if the latter.
+
+While this is parsimonious with respect to syntax, it creates more complexity for static analyzers.
+It is currently statically known that the `expr` in `expr.${idx}` must be a set and that `idx` must be a string.
+Representing a new constraint that _either_ `expr` is a set and `idx` is a string _or_ `expr` is a list and `idx` is an integer might be difficult for such tools and could result in a degraded user experience.
+
+#### `expr[1]`
+
+The most common syntax, by far, for indexing into a list or array in other programming languages is `expr[idx]`.
+Nix is currently whitespace-insensitive with respect to attrpaths, so such an expression is indistinguishable from `expr [idx]`, an application of the function `expr` to the argument `[idx]`.
+It would be possible to adapt the Nix lexer to interpret the `[` character in `expr[idx]` as the opening of an indexer or a list depending on whether there is whitespace before it.
+Among the disadvantages of this approach would be that it would make specifying the grammar of Nix more complicated, and that it may prevent long attrpaths from being broken naturally across lines.
+Perhaps most fatally, despite the odds of someone writing `expr[idx]` and intending a function call being virtually nil, the conservative principles of the Nix team forbid altering the meaning of even such an unlikely bit of syntax without some sort of larger language versioning or deprecation story, which has eluded us for some time.
+
+All of the above objections apply to `expr(1)` as well, with the additional drawback that in a wide array of C-like languages, this syntax represents exactly what it currently (if coincidentally) represents in Nix.
+
+#### `expr.$[2]`
+
+The syntax `expr.$[idx]` was offered as a possibility in [NixOS/nix#10949](https://github.com/NixOS/nix/issues/10949).
+It resembles the `${}` syntax already used in attrpaths, with the change in delimiters suggesting a shift from attribute sets to lists.
+However, it requires an additional character to type and its technical qualities are identical to those of the proposed syntax without the `$` character.
+There is at least some prior art for `.[]` in OCaml and F#; there is none that I know of for `.$[]`.
+
+#### `expr.3`
```

Comment:
Everyone arguing for this alternative: you don't have to convince me of its technical superiority. I am already convinced, as I describe below. This is the syntax I'm using in my personal Nix build.

Rather, you have to convince me that the Nix maintainers can be persuaded to make a breaking change to float literal syntax on a timeline that is less than, say, five years, when they have expressed reluctance to make ‘breaking changes’ to the language even with respect to [obvious](https://github.com/NixOS/nix/pull/10025) buggy [behavior](https://github.com/NixOS/nix/pull/11121).

Gating this proposal behind something that will never happen is, in practice, equivalent to rejecting the proposal. I would rather see *some* syntax implemented in a couple of years than wait indefinitely for the *ideal* syntax.

And if sane language evolution manages to actually become policy at some point, in that case we'll have the tools we need to deprecate the less-appealing syntax and migrate to the better one, should we choose to do so.

Reply:
With `expr.3` how would be non-literal index look like?

Obviously, `expr.n` is already mean `expr.${"n"}`.

If you introduce `expr.(n)` or `expr.[n]` for non-literal index, should they work with literals, should `expr.(3)` or `expr.[3]` be valid too?

If yes, why duplicating and confusing `expr.3` ?

Reply:
This proposal is only for when the desired index is known; replacing `elemAt` for variable indices is left as [future work](https://github.com/rhendric/nix-rfcs/blob/rhendric/list-index-syntax/rfcs/0181-list-index-syntax.md#future-work).

Reply:
I believe that adopting `expr.3` without addressing the implications for non-literal indices could lead to significant challenges down the road. If we postpone even the discussion on non-literal indices, we risk creating a legacy issue that will be difficult and cumbersome to address in the future.

While we might agree on `expr.3` today, this decision could result in a syntax that feels outdated and problematic later on, making it harder to evolve the language cleanly.

Reply:
I think I agree here. If we agree that `elemAt` is the future then that is fine. But I think if we agree that we want to improve the dynamic syntax in the future it is best to have that discussion now to ensure that we don't shoot ourselves in the foot.

Reply:
There are several options:
* `elemAt` is good enough
* Make `.${n}` work with both numbers on lists as well as strings on sets
* Introduce `.$[n]` as a parallel construction to `.${str}`
* If this proposal is adopted as-is, simply extend `.[n]` to also allow expressions

I'm confident that regardless of whether this proposal is adopted as-is or if the `expr.3` alternative is chosen instead, there are non-problematic ways forward if we *do* agree that `elemAt` doesn't suffice (and I don't think we do; at least, I'm much less concerned about that personally).

Reply:
Also, besides the space issue with float, `expr.3` opens a gate to hard to read space issues like `expr.3a` (which is a function application `expr.3 a`) or `expr.1e+3`

Reply:
It doesn't really open that gate; `3a` and `1e+3` are already ill-advised things to write in Nix.

Reply:
And for floats, even if we forbid the `.3` syntax for floats, how this would be parsed?
`let a0=1; in [3a0.3a0]`

`[3 a 0.3 a0]` or `[3 a0.$[3] a0]`

Reply:
> * Make `.${n}` work with both numbers on lists as well as strings on sets
this is already criticized above for runtime overhead and issues with static analyzers.

> * Introduce `.$[n]` as a parallel construction to `.${str}`
> * If this proposal is adopted as-is, simply extend `.[n]` to also allow expressions

These are not truly parallel, `[]` is a bit different in Nix than `{}` and `[]`: while `${a b}` and `(a b)` are function applications, `[a b]` is not.

How to read `.$[a b]` or `.[a b]` then?

Reply:
I would add `expr.(3)` or `expr.(n)` variant to the proposal, it seems to be the less confusing.

Reply:
> `3a` and `1e+3` are already ill-advised things to write in Nix.

I mean, currently, such code likely hit "attempt to call something which is not a function but an integer" error
but with `expr.3` one could mistakenly write `expr.0x29A` and she may get no error even at runtime.
Just wrong result at the end.

Reply:
I would add `expr.(3)` or `expr.(n)` variant to the proposal, it seems to be the less confusing, for

1. no issues with floats
2. no issues with some letter or symbol glued on the right side
3. no issues with sets: `.(a)` and `.("a")` are already forbidden to use with sets; really forbidden, not just considered ill-formed, there is guarantied no such existing Nix code
4. no issues with spaces: `expr.(a)` is `expr .(a)` is `expr. (a)` is `expr . ( a )`
5. unlike `[]`, priorities inside `()` are surprise-free: `expr.(a b)` means application like in `expr.${a b}`

Reply:
> Rather, you have to convince me that the Nix maintainers can be persuaded to make a breaking change to float literal syntax on a timeline that is less than, say, five years

At least for Lix, I am currently setting up [infrastructure](https://gerrit.lix.systems/c/lix/+/1735?tab=comments) to [remove URL literals](https://gerrit.lix.systems/c/lix/+/1736), and the `.3` floating literals are guaranteed to be next.

However, I can't guarantee that adding the new syntax will be allowed without proper language versioning tools (which are on the roadmap but it will take years still).

> elemAt is good enough

I'll join that camp. Especially since the main motivation for this RFC is the REPL, where the user knows which index they want to use.

Reply:
(What other projects do is their business; this is a Nix RFC.)

For everyone using original Nix, a reminder that you can have either the proposed syntax or the `expr.3` alternative in your REPL [right now](https://gitlab.com/rhendric/nix-list-index-syntax), on any stable or unstable version of Nix, without waiting for anyone else to make any decisions about what is allowed.